### PR TITLE
[IMP] mail: introduce unread counter

### PR DESCRIPTION
**Pull request by phenix-factory**

### CODE REVIEW

**Comment by alexkuhn** [comment]
LGTM, would be nice to validate the change with test. Likely some tests that can be unskipped

**Comment by phenix-factory** [comment]
@alexkuhn To my surprise, there was no specific test that was targeting this feature. Or I couldn't find it.
So I added one.

File: addons/mail/static/src/new/thread/message_service.js [hunk 1]
```diff
@@ -431,8 +438,12 @@ export class MessageService {
         group.resIds.add(data.resId);
     }
 
+    /**
+     * @param {import("@mail/new/core/thread_model").Thread} thread
+     * @returns {import("@mail/new/core/thread_model").Thread["messages"]}
+     */
     sortMessages(thread) {
-        thread.messages.sort((msgId1, msgId2) => {
+        return thread.messages.sort((msgId1, msgId2) => {
```
Comment (alexkuhn) [inline addons/mail/static/src/new/thread/message_service.js]: Don't rely on return. This method sort the thread in place, just inspect `thread.messages` after calling it.

File: addons/mail/static/src/new/thread/thread_service.js [hunk 2]
```diff
@@ -601,6 +608,68 @@ export class ThreadService {
...
+        let baseCounter = thread.serverMessageUnreadCounter;
+        let countFromId = thread.serverLastSeenMsgBySelf ? thread.serverLastSeenMsgBySelf : 0;
+        const orderedMessages = this.message
+            .sortMessages(thread)
+            .map((messageId) => this.store.messages[messageId]);
+        // But if the client knows the last seen message that the server
+        // returned (and by assumption all the messages that come after),
+        // the counter can be computed fully locally, ignoring potentially
+        // obsolete values from the server.
+        const lastSeenByCurrentPartnerMessageId = this.lastSeenByCurrentPartnerMessageId(
```
Comment (alexkuhn) [inline addons/mail/static/src/new/thread/thread_service.js]: Replace `CurrentPartner` to `Self`. This reduces noise a bit, which improves readability.

File: addons/mail/static/src/new/thread/thread_service.js [hunk 3]
```diff
@@ -601,6 +608,68 @@ export class ThreadService {
...
+            lastSeenByCurrentPartnerMessageId >= firstMessage.id
+        ) {
+            baseCounter = 0;
+            countFromId = lastSeenByCurrentPartnerMessageId;
+        }
+        // Include all the messages that are known locally but the server
+        // didn't take into account.
+        return orderedMessages.reduce((total, message) => {
+            if (message.id <= countFromId) {
+                return total;
```
Comment (alexkuhn) [inline addons/mail/static/src/new/thread/thread_service.js]: I prefer to remove all comments. It's not that useful, and it just adds noise. Code is easier to ready without them.

**Pull request merged**